lab_4/main.py

```diff
@@ -1,56 +1,158 @@
+    tokens = []
+    for sentence in sentences:
+        list_tokens = re.sub('[^a-z \n]', '', sentence.lower()).split()
+        if not list_tokens:
+            continue
+        tokens += list_tokens + ['<END>']
+    return tuple(tokens)
 
 
 class WordStorage:
     def __init__(self):
-        pass
+        self.storage = {}
 
     def _put_word(self, word: str):
-        pass
+        if not isinstance(word, str) or not word:
+            raise ValueError
+
+        if word not in self.storage:
+            self.storage[word] = len(self.storage) + 1
+        return len(self.storage)
 
     def get_id(self, word: str) -> int:
-        pass
+        if not isinstance(word, str) or not word:
+            raise ValueError
+
+        if word not in self.storage:
+            raise KeyError
+        return self.storage[word]
 
     def get_word(self, word_id: int) -> str:
-        pass
+        if not isinstance(word_id, int):
+            raise ValueError
+
+        if word_id not in self.storage.values():
+            raise KeyError
+        for key, value in self.storage.items():
+            if value == word_id:
+                return key
 
     def update(self, corpus: tuple):
-        pass
+        if not isinstance(corpus, tuple) or not tuple:
+            raise ValueError
+
+        for word in corpus:
+            self._put_word(word)
 
 
 def encode_text(storage: WordStorage, text: tuple) -> tuple:
-    pass
+    if not isinstance(text, tuple) or not all(isinstance(word, str) for word in text) \
+            or not isinstance(storage, WordStorage):
+        raise ValueError
+
+    encoded_text = []
+    for word in text:
+        encoded_text.append(storage.get_id(word))
+    return tuple(encoded_text)
 
 
 class NGramTextGenerator:
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie):
-        pass
+        self._word_storage = word_storage
+        self._n_gram_trie = n_gram_trie
+
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or not all(isinstance(num, int) for num in context)\
+                or self._n_gram_trie.size != len(context) + 1:
+            raise ValueError
+
+        word_freq = 0
+        word_ind = 0
+        for n_gram, freq in self._n_gram_trie.n_gram_frequencies.items():
+            if n_gram[:len(context)] == context and freq > word_freq:
+                word_freq = freq
+                word_ind = n_gram[-1]
+        if not word_freq:
+            max_freq_word = max(self._n_gram_trie.uni_grams.values())
+            return list(word[0] for word, value in self._n_gram_trie.uni_grams.items() if value == max_freq_word)[0]
+        return word_ind
+
 
     def _generate_sentence(self, context: tuple) -> tuple:
-        pass
+        if not isinstance(context, tuple) or not all(isinstance(num, int) for num in context):
+            raise ValueError
+
+        sentence = list(context)
+        for counter in range(20):
```

Comment:
If you do not intend to use the var from `for`-loop, then it is commonly named as `_`.